content/en/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions.md

```diff
@@ -994,6 +996,131 @@ Here is the declarations type mapping between OpenAPIv3 and CEL type:
+| If previous value was X, new value can only be A or B, not Y or Z | `oldSelf != 'X' \|\| self in ['A', 'B']`
+| Nondecreasing counters                                            | `self >= oldSelf`
+
+#### Resource Constraints
+
+Resource consumption of CEL expressions is checked at CustomResourceDefinition apply time. If an expression is
+estimated to be prohibitively expensive to execute, it will result in a validation error. A similar 
+system is used at runtime that observes the actions the CEL interpreter takes. If the interpreter executes
+too many instructions, execution of the expression will be halted, and an error will result.
+Each CustomResourceDefinition is also allowed a certain amount of resources to finish executing all of 
+its CEL expressions. If the sum total of its expressions are estimated at apply time to go over that limit, 
+then a validation error will also occur.
+
+It is considered best practice to put a limit via `maxItems`, `maxProperties`, and `maxLength` for anything
+that will be processed in a CEL rule in order to prevent validation errors during cost estimation. For 
+example, given this schema with one rule:
+
+```yaml
+openAPIV3Schema:
+  type: object
+  properties:
+    foo:
+      type: array
+      items:
+          type: string
+      x-kubernetes-validations:
+        - rule: "self.all(x, x.contains('a string'))"
+```
+
+The cost system will not allow this rule. Without limits being set, the estimated cost of this rule will
+exceed the per-expression cost limit. But if we add limits in the appropriate places, the rule will be allowed:
+
+```yaml
+openAPIV3Schema:
+  type: object
+  properties:
+    foo:
+      type: array
+      maxItems: 25
+      items:
+        type: string
+        maxLength: 10
+      x-kubernetes-validations:
+        - rule: "self.all(x, x.contains('a string'))"
+```
+
+The cost estimation system takes into account how many times the rule will be executed in addition to the
+data that the rule is working with. For instance, the following rule will have the same estimated cost as the
+previous example (despite the rule now being defined on the individual array items):
+
+```yaml
+openAPIV3Schema:
+  type: object
+  properties:
+    foo:
+      type: array
+      maxItems: 25
+      items:
+        type: string
+        x-kubernetes-validations:
+          - rule: "self.contains('a string'))"
+        maxLength: 10
+```
+
+If a list inside of a list has a CEL expression that uses `self.all`, that is significantly more expensive 
+than a non-nested list with the same expression. A rule that would have been allowed on a non-nested list might need lower limits set on both nested lists in order to be allowed. Nested traversals on the same
+list (i.e, nesting calls to `self.all` or similar) should be avoided where possible for similar reasons. 
+
+In general, both systems will allow expressions that do not need to iterate; these expressions will
```

Comment:
s/expressions/rules?  Best to pick a single term and use it uniformly. (this suggestion applies to this entire PR).